pkg/migration/dryrun.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	// We don't block migration if difference is found in spec. Instead
+	// raise an event and let user review before approving migration.
+	if err := h.validateSpec(cluster, ds); err != nil {
+		k8sutil.WarningEvent(h.ctrl.GetEventRecorder(), cluster, util.DryRunFailedReason,
+			fmt.Sprintf("Spec validation failed: %v", err))
+	} else {
+		k8sutil.InfoEvent(h.ctrl.GetEventRecorder(), cluster, util.DryRunCompletedReason,
+			"Spec validation completed successfully")
+	}
+
+	return nil
+}
+
+func (h *Handler) validateSpec(cluster *corev1.StorageCluster, ds *appsv1.DaemonSet) error {
+	var msg string
+	pod, err := h.ctrl.CreatePodTemplate(cluster, &v1.Node{}, "")
+	if err != nil {
+		return err
+	}
+
+	err = h.deepEqualPod(&ds.Spec.Template, &pod)
+	if err != nil {
+		msg += fmt.Sprintf("Validate portworx pod failed, %v\n", err)
+	}
+
+	if msg != "" {
+		return fmt.Errorf(msg)
+	}
+	return nil
+}
+
+// DeepEqualPod compares two pods.
+func (h *Handler) deepEqualPod(p1, p2 *v1.PodTemplateSpec) error {
+	var msg string
+
+	err := h.deepEqualVolumes(p1.Spec.Volumes, p2.Spec.Volumes)
+	if err != nil {
+		msg += fmt.Sprintf("Validate volumes failed: %s\n", err.Error())
+	}
+
+	err = h.deepEqualContainers(p1.Spec.Containers, p2.Spec.Containers)
+	if err != nil {
+		msg += fmt.Sprintf("Validate containers failed: %s\n", err.Error())
+	}
+
+	err = h.deepEqualContainers(p1.Spec.InitContainers, p2.Spec.InitContainers)
+	if err != nil {
+		msg += fmt.Sprintf("Validate init-containers failed: %s\n", err.Error())
+	}
+
+	if !reflect.DeepEqual(p1.Spec.ImagePullSecrets, p2.Spec.ImagePullSecrets) {
+		msg += fmt.Sprintf("ImagePullSecrets are different: first %+v, second: %+v\n", p1.Spec.ImagePullSecrets, p2.Spec.ImagePullSecrets)
+	}
+
+	if msg != "" {
+		return fmt.Errorf(msg)
+	}
+	return nil
+}
+
+func (h *Handler) deepEqualVolumes(vol1, vol2 []v1.Volume) error {
+	// Use host path as key, as volume name may be different but path is the same.
+	getVolumeKey := func(v interface{}) string {
+		vol := v.(v1.Volume)
+
+		if vol.HostPath != nil {
+			return vol.HostPath.Path
+		}
+
+		return vol.Name
+	}
+
+	getVolumeList := func(arr []v1.Volume) []interface{} {
+		var objs []interface{}
+		for _, obj := range arr {
+			// "crioconf" only exists with operator install.
+			// "dev" only exists with daemonset install.
+			if obj.Name == "crioconf" || obj.Name == "dev" {
+				continue
+			}
+			objs = append(objs, obj)
+		}
+		return objs
+	}
+
+	return util.DeepEqualObjects(
+		getVolumeList(vol1),
+		getVolumeList(vol2),
+		getVolumeKey,
+		h.deepEqualVolume)
+}
+
+func (h *Handler) deepEqualVolume(obj1, obj2 interface{}) error {
+	t1 := obj1.(v1.Volume)
+	t2 := obj2.(v1.Volume)
+
+	vol1 := t1.DeepCopy()
+	vol2 := t2.DeepCopy()
+
+	cleanupFields := func(vol v1.Volume) v1.Volume {
+		// We dont compare name, but only the content
+		// Ignore hostPath type as nil may represent default value.
+		vol.Name = ""
+		if vol.HostPath != nil {
```

Comment:
I don't think there is `*` type in hostpath. The support types are here - https://kubernetes.io/docs/concepts/storage/volumes/#hostpath.

Also, we have some volumes that use a different type like `DirectoryOrCreate`. It would be good to validate that as well.

Reply:
* was caught during test. the path is still validated even for DirectoryOrCreate type, is it sufficient?